test/test_images/internal/knockdown/knockdown.go

```diff
@@ -39,92 +44,84 @@ type Config struct {
 	Time time.Duration `envconfig:"TIME" required:"true"`
 }
 
+type Receiver interface {
+	Knockdown(event cloudevents.Event) bool
+}
+
 // Main should be called by the process' main function. It will run the knockdown test. The return
 // value MUST be used in os.Exit().
 func Main(config Config, kdr Receiver) int {
-	// Note that this only accepts TraceContext style tracing, not B3 style tracing. This is being
-	// done here in test code so that we are effectively asserting that our components output in
-	// TraceContext format. For all production code, kncloudevents.NewDefaultClient() should be used
-	// instead.
-	client, err := cloudevents.NewDefaultClient()
+	p, err := http.New()
 	if err != nil {
-		panic(err)
-	}
-
-	r := receiver{
-		kdr: kdr,
-	}
-
-	var ctx context.Context
-	ctx, r.cancel = context.WithCancel(context.Background())
-
-	timer := time.NewTimer(config.Time)
-	defer timer.Stop()
-	go func() {
-		<-timer.C
-		// Write the termination message if time out
-		fmt.Printf("Time out waiting for the knock down event(s).")
-		if err := r.writeFailedTerminationMessage(); err != nil {
-			fmt.Printf("Failed to write termination message, %v\n", err)
-		}
-		r.cancel()
-	}()
-
-	fmt.Printf("Waiting to receive event (timeout in %s)...", config.Time.String())
-
-	// Note, we assume that closing the context will gracefully shutdown the receiver. Having looked
-	// at the current implementation, this is true. But nothing guarantees it in the future.
-	if err := client.StartReceiver(ctx, r.Receive); err != nil {
 		log.Fatal(err)
 	}
 
-	return 0
-}
-
-type receiver struct {
-	kdr Receiver
-
-	// cancel will cancel the main receiver's context. Calling this will cause Main() to become
-	// unblocked and call os.Exit(0). In general, write{Successful,Failed}TerminationMessage should be
-	// called first.
-	cancel func()
-}
-
-type Receiver interface {
-	Knockdown(event cloudevents.Event) bool
-}
-
-func (r *receiver) Receive(event cloudevents.Event) {
-	done := r.kdr.Knockdown(event)
-	if done {
-		if err := r.writeSuccessfulTerminationMessage(); err != nil {
-			fmt.Printf("Failed to write termination message, %s.\n", err.Error())
+	fmt.Printf("Waiting to receive event (timeout in %v)...\n", config.Time)
+	eg, ctx := errgroup.WithContext(context.Background())
+	pctx, pcancel := context.WithTimeout(ctx, config.Time)
+	ctx, cancel := context.WithCancel(ctx)
+	eg.Go(func() error {
+		defer cancel()
+		return p.OpenInbound(pctx)
+	})
+	eg.Go(func() error {
+		var done bool
+		for {
+			msg, err := p.Receive(ctx)
+			if err == io.EOF {
+				if done {
+					return nil
+				}
+				return errors.New("knockdown not done")
+			}
+			if err != nil {
+				return err
+			}
+			e, err := binding.ToEvent(ctx, msg)
+			if err != nil {
+				msg.Finish(err)
+				return err
+			}
+			if kdr.Knockdown(*e) {
+				done = true
```

Comment:
Why set `done` here, rather than:
```golang
pcancel()
msg.Finish(nil)
return nil
```

Reply:
The receiver can't return here as the server may accept additional requests before shutting down. In order for the server to shutdown gracefully those requests need to be processed so the receiver will continue running until it gets an EOF.

Reply:
Please add a comment to that affect. Essentially stating why we need the `done` variable.